mars-sim-core/src/main/java/com/mars_sim/core/unit/AbstractMobileUnit.java

```diff
@@ -0,0 +1,245 @@
+		return unitManager.getBuildingByID(currentBuildingInt);
+	}
+
+	/**
+	 * Computes the building the person is currently located at.
+	 * Returns null if outside of a settlement.
+	 *
+	 * @return building
+	 */
+	public void setCurrentBuilding(Building building) {
+		if (building == null) {
+			currentBuildingInt = -1;
+		}
+		else {
+			currentBuildingInt = building.getIdentifier();
+		}
+	}
+
+
+	/**
+	 * Gets the unit's location.
+	 *
+	 * @return the unit's location
+	 */
+	@Override
+	public Coordinates getCoordinates() {
+		Unit cu = getContainerUnit();
+		if (cu.getUnitType() == UnitType.MARS) {	
+			// Since Mars surface has no coordinates, 
+			// Get from its previously setting location
+			return super.getCoordinates();
+		}
+		
+		// Unless it's on Mars surface, get its container unit's coordinates
+		return cu.getCoordinates();
+	}
+
+    /**
+	 * This method assumes the Unit could be movable and change container. It identifies the
+	 * appropriate container it can use. Ideally this method should be moved to a 
+	 * new subclass called 'MovableUnit' that encapsulates some positioning methods 
+	 * not applicable to Structures.
+	 */
+	public String getContext() {
+		// Check vehicle first because could be in a Vehicle in a Settlement
+		if (isInVehicle()) {
+			return getVehicle().getChildContext();
+		}
+		else if (isInSettlement()) {
+			var b = getBuildingLocation();
+			if (b != null) {
+				return b.getContext();
+			}
+			else {
+				return getAssociatedSettlement().getName();
+			}
+		}
+		else if (isOutside()) {
+			return getCoordinates().getFormattedString();
+		}
+		else {
+			return getContainerUnit().getName();
+		}
+	}
+
+    /**
+     * Settlement that is associated with this Unit.
+     */
+    @Override
+    public Settlement getAssociatedSettlement() {
+        return owner;
+    }   
+    
+    /**
+	 * Gets the settlement the Unit is at.
+	 * Returns null if unit is not at a settlement.
+	 *
+	 * @return the settlement
+	 */
+	public Settlement getSettlement() {
+
+		if (getContainerID() <= Unit.MARS_SURFACE_UNIT_ID)
+			return null;
+
+		var c = getContainerUnit();
+
+		if (c instanceof Settlement s) {
+			return s;
+		}
+
+		else if (c instanceof Vehicle v) {
+			// Will see if vehicle is inside a garage or not
+			return (v.isInVehicleInGarage() ? v.getSettlement() : null);
+		}
+
+		else if (c instanceof FixedUnit b) {
+			return b.getAssociatedSettlement();
+		}
+		else if (c instanceof AbstractMobileUnit m) {
+			return m.getSettlement();
+		}
+
+		return null;
+	}
+
+    /**
+	 * Is this unit inside a settlement ?
+	 *
+	 * @return true if the unit is inside a settlement
+	 */
+	@Override
+	public boolean isInSettlement() {
+		return (getSettlement() != null);
+	}
+
+    
+	/**
+	 * Gets vehicle person is in, null if person is not in vehicle.
+	 *
+	 * @return the person's vehicle
+	 */
+	@Override
+	public Vehicle getVehicle() {
+		if (getLocationStateType() == LocationStateType.INSIDE_VEHICLE) {
```

Comment:
So, will you be removing the need of calling LocationStateType next and strictly using container unit for location determination ?

Reply:
Not sure yet, it would be good if we could but taking this slowly. I have created a new project to cover all related chabges